### PR TITLE
Fixed imbuement system behavior and a crash related to memory reference

### PR DESCRIPTION
**Pull request by hyresu**
# Description

The imbuements are not behaving as expected

### **Actual**

- Imbue an item inside a backpack and you get a permanent imbuement bonus until relog
- When clearing an imbuement from an item you don't lose the imbuement
- You can't use bank balance for imbuements
- Elemental damage from weapons is not working if item is imbued
- Critical hit chance goes up to 20%
- Falcon coif and Dark Whispers can not be imbued with magic level
- Item description shows crit chance with positive sign and "critical hit damage"

### **Expected**

- Item only gets inmediate bonus when imbued on equipment slots
- Clearing the imbuement will make you lose it if the item is equipped
- You should be able to use bank balance for imbuements
- Elemental damage shouldn't be affected
- Critical hit chance should not get higher than 10% if the item already has critical hit chance
- Falcon Coif should be able to get Intricate Epiphany and Dark Whispers only Basic Epiphany
- Item description should show crit chance without positive sign and "critical extra damage"
- Move map deletion to outside the loop to avoid memory reference crash on Game::checkImbuements function
- Where the "imbuementslot" value is from 1 to 3 and the imbuement name values are also from 1 to 3, which is the imbuement level (basic, intricate or powerful), the name must be placed as above, as it is from the name that the value of each imbuement is called by ItemParse.

## Fixes

All of the mentioned behaviours above

## Type of change

Please delete options that are not relevant.

  - [x] Bug fix (non-breaking change which fixes an issue)

## Checklist

  - [x] My code follows the style guidelines of this project
  - [x] I have performed a self-review of my own code
  - [x] I checked the PR checks reports
  - [x] I have commented my code, particularly in hard-to-understand areas
  - [x] I have made corresponding changes to the documentation
  - [x] My changes generate no new warnings
  - [x] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: src/items/item.h [hunk 1]
```diff
@@ -1144,8 +1144,12 @@ class Item : virtual public Thing
 		 */
 		bool getImbuementInfo(uint8_t slot, ImbuementInfo *imbuementInfo);
 		void setImbuement(uint8_t slot, uint16_t id, int32_t duration);
-		bool hasImbuementType(ImbuementTypes_t imbuementType) {
-			return items[id].imbuementTypes.find(imbuementType) != items[id].imbuementTypes.end();
+		bool hasImbuementType(ImbuementTypes_t imbuementType, uint16_t maxTier) {
```
Comment (lgrossi) [inline src/items/item.h]: Why do you need maxTier here?

File: src/creatures/players/player.cpp [hunk 2]
```diff
@@ -5428,12 +5434,12 @@ uint16_t Player::getFreeBackpackSlots() const
...
 }
 
-void Player::addItemImbuementStats(const Imbuement* imbuement)
+void Player::addItemImbuementStats(const Imbuement* imbuement, const ItemType& it)
 {
 	bool requestUpdate = false;
 	// Check imbuement skills
 	for (int32_t i = SKILL_FIRST; i <= SKILL_LAST; ++i) {
-		if (imbuement->skills[i]) {
+		if (imbuement->skills[i] && (i != SKILL_CRITICAL_HIT_CHANCE || !it.abilities || !it.abilities->skills[i])) {
```
Comment (lgrossi) [inline src/creatures/players/player.cpp]: this is a bit too hacky, would love to see an explanation of the need to come up with a better solution or at least a nicer if statement.

File: src/creatures/players/player.cpp [hunk 3]
```diff
@@ -5471,13 +5477,13 @@ void Player::addItemImbuementStats(const Imbuement* imbuement)
 	return;
 }
 
-void Player::removeItemImbuementStats(const Imbuement* imbuement)
+void Player::removeItemImbuementStats(const Imbuement* imbuement, const ItemType& it)
```
Comment (lgrossi) [inline src/creatures/players/player.cpp]: same

**Comment by lgrossi** [comment]
Nice man, good to see some contributions appearing. Good job, the PR is neat. We actually don't know much about the imbuement system, how it works in global, so we might have missed some details, thanks for correcting.

The only thing is that some extra checks were not explained, I'd love to understand them to see if that would be the best approach.

Thanks again!

**Review by dudantas — APPROVED** [review]

**Pull request merged**